pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UnusedPrivateMethodRule.java

```diff
@@ -51,29 +52,38 @@ protected Collection<String> defaultSuppressionAnnotations() {
 
     @Override
     public Object visit(ASTCompilationUnit file, Object param) {
-        // this does a couple of traversals:
-        // - one to find annotations that potentially reference a method
-        // - one to collect candidates, that is, potentially unused methods
-        // - one to walk through all possible usages of methods, and delete used methods from the set
-        Set<String> methodsUsedByAnnotations = methodsUsedByAnnotations(file);
-        Map<JExecutableSymbol, ASTMethodDeclaration> candidates = findCandidates(file, methodsUsedByAnnotations);
+        return addViolations(file, asCtx(param));
+    }
 
+    /**
+     * @implNote does following traversals:
+     * <ol>
+     * <li>find annotations that potentially reference a method</li>
+     * <li>collect candidates, that is, potentially unused methods</li>
+     * <li>walk through all possible usages of methods, and delete used methods from the set</li>
+     * </ol>
+     */
+    private Map<JExecutableSymbol, ASTMethodDeclaration> findViolations(ASTCompilationUnit file,
+                                                                        Map<JExecutableSymbol, ASTMethodDeclaration> candidates) {
         file.descendants()
             .crossFindBoundaries()
             .<MethodUsage>map(asInstanceOf(ASTMethodCall.class, ASTMethodReference.class))
-            .forEach(ref -> {
```

Comment:
Every indentation serves a dedicated purpose, making it a suitable candidate for the separation of concerns pattern.